### ISSUE TITLE
NuGet.VisualStudioExtension: NuGet automatically adds "packages" to source control (TFS)

### ISSUE THREAD

**Issue by bert2** [NuGet/Home#464]
**tl;dr:** When installing new packages, NuGet shouldn't add the content of "packages" to the pending changes by default.

**Repro steps:**
* In Visual Studio 2013, create a new solution with a class library project.
* Add the solution to TFS.
* Install some packages with NuGet.

**Expected:**
* Only the project/solution files along with the "packages.config" file should be added to source control.
* NuGet should not automatically add the contents of the "packages" folder to source control (for the reasons why, see the introductory text in this article about [Package Restore from the NuGet docs](https://docs.nuget.org/Consume/Package-Restore)).
* NuGet should offer a user-configurable option controlling this behaviour.

**Actual:**
* After the last repro step, NuGet also added the contents of the "packages" folder to the pending changes list.
* In the Visual Studio options there is no option to prevent those unwanted pending changes.
* Users have to resort to work-arounds in order to prevent this behaviour (either clean the pending changes manually or [mix <2.7 Packe Restore with >2.7 Package Restore](https://nuget.codeplex.com/discussions/472174)).

**Comment by csharpfritz** [comment]
There is a configuration option provided in NuGet.config to disable source
control integration and it is documented here:
https://docs.nuget.org/Consume/package-restore/team-build

However, I agree that we need a more user-friendly and discoverable way to
enable this behavior

**Comment by bert2** [comment]
@csharpfritz after the last repro step there is no "NuGet.config" in my solution directory. And -- to be blunt -- I dont want to create one either. I want to use "Automatic Package Restore" [which is supposed to give me the following advantages](https://docs.nuget.org/Consume/Package-Restore):

> * **No need to enable it for your project or solution**. Visual Studio will automatically download missing packages before your projects are built and team members don't need to understand NuGet Package Restore.
> * **No additional directories and files are required within your project or solution**.

However, I tried your work-around and it did prevent NuGet from adding the content of "packages" to the pending changes. They are only detected as regular adds by TFS now.

**Comment by yishaigalatzer** [comment]
@bert2 you don't have to put this in your solution, this could be in your userprofile nuget.config

**Comment by bert2** [comment]
@yishaigalatzer Interesting, I did not know that. However this is not an option in my case, since it requires new developers to know about this setting and have it activated in their user-profile.

**Comment by yishaigalatzer** [comment]
Then leave it in the solution. What is broken about it?

**Comment by bert2** [comment]
I consider it broken, because the standard behaviour contradicts the [NuGet documentation](https://docs.nuget.org/Consume/Package-Restore). That aside, adding binaries to source control is simply bad practice. NuGet may offer to opt-in for source-controlled libraries, but it shouldn't force the user to resort to work-arounds with outdated configuration switches (using nuget.config is the pre 2.7 way of package restore).

**Comment by csharpfritz** [comment]
The documentation we linked, and our replies agree - its not a
user-friendly solution, it is a work-around.  NuGet.config is always where
all of your NuGet repositories are listed and is searched hierarchically.

Agreed, adding binaries to source-control for DEV branches is a bad
practice.  Adding binaries to a master or production quality branch I
consider a very good practice so that you always have all dependencies
ready to deploy from a clean production instance of code.

**Comment by bert2** [comment]
I just realized that there already is a similar report in [the old CodePlex issue tracker](https://nuget.codeplex.com/workitem/4072) (@csharpfritz's link lead me to it) which has been updated today by @feiling. The proposed version for the fix is 3.0.

**Comment by bert2** [comment]
@csharpfritz, so we agree that NuGet shouldn't add the content of "packages" to the pending changes by default?

**Comment by yishaigalatzer** [comment]
This is a debate about the default behavior. We are not changing it for now. The reason it is done this way is that we unfortunately cannot auto opt you into downloading packages.

Will consider changing this in future releases.

I do understand you dislike it, but any default we choose will break someone's flow.

**Comment by bert2** [comment]
This is not only a debate about default behaviour. Its about behaviour the documentation implies, but NuGet doesn't exhibit.
In which way you prefer to fix this discrepancy -- either update the docs or change the behaviour of NuGet -- is of course up to you. But please fix it. I, and propably other new NuGet users as well, will much appreciate a less confusing user experience regarding this issue.

In my eyes, the optimal compromise would be to have a config switch controlling this behaviour accessible through the Visual Studio options.

**Comment by deepakaravindr** [comment]
This doc might throw some light: http://docs.nuget.org/consume/package-restore#omitting-packages-from-source-control

**Comment by bert2** [comment]
Actually, it reveals more odd things. If I followed the proposed ways to omit packages from source control for both Git and TFS I'd get two different results. Git would completely ignore my packages whereas TFS would not have them in the pending changes but still detect them as adds. So the user actually would have to add the ".tfignore" file as well. But why isn't it sufficient to only use ".tfignore" in order to achieve the same thing I can achieve with ".gitignore"?

Also, when you read the document from the top to bottom it appears to lie a bit. It starts out with arguing _against_ adding packages to source control. Then it explains the things Automatic Package Restore does better compared to older versions: it's activated by default, less intrusive, and _requires no additional files_.
I'm like: yes, that's totally what I need and want! But then I've hit the section "Omitting Packages from Source Control" and went from "Yaaay!" to "Wait... what?" Why is this even there? From all I've read above, shouldn't it say "Forcing NuGet to Add Packages to Source Control"?

**Comment by Gladskih** [comment]
In current state new automatic package restore is unusable. Technologies like NuGet is designed to simplify dependency handling, but now it leads to time wasting for googling, reading forums, mess configuring.

**Comment by yishaigalatzer** [comment]
@gladskih please be more specific about your pain. We can't do anything with a general statement.

**Comment by deepakaravindr** [comment]
This is a duplicate of https://github.com/NuGet/Home/issues/493
